src/liballoc/raw_vec.rs

```diff
@@ -235,7 +235,66 @@ impl<T, A: Alloc> RawVec<T, A> {
         }
     }
 
-    /// Doubles the size of the type's backing allocation. This is common enough
+    /// Grows the vector capacity by `capacity_increase`.
+    ///
+    /// It allows implementing amortized O(1) `push` on vector-like containers.
+    ///
+    /// # Attributes
+    ///
+    /// - `#[inline]`: LLVM is able to completely elide memory allocations in
+    ///   many cases if it can "see" where memory is allocated and freed.
+    /// - `#[cold]`: calling this function is a "rare" event
+    ///
+    #[inline]
+    #[cold]
+    pub fn grow_by(&mut self, capacity_increase: usize) {
+        let elem_size = mem::size_of::<T>();
+        assert!(elem_size != 0, "RawVecs of zero-sized types can't grow");
+
+        let (new_cap, uniq) = match self.current_layout() {
+            Some(cur) => {
+                // The invariant `elem_size * self.cap <= isize::MAX` is
+                // maintained by `alloc_guard`; the alignment will never be too
+                // large as to "not be specifiable" (so we can use
+                // `from_size_align_unchecked`).
+                let new_cap = Self::suitable_capacity(self.cap, capacity_increase);
```

Comment:
A wraparound/overflow check is needed here, or in that method, or somewhere close. Or an argument why it's not needed.

Reply:
(The old code used the argument that the new capacity was <= 2 * old_capacity)

Reply:
The new capacity was always `2 * old_capacity`.

Reply:
I'll check this. I am adding some tests close/at the `i32::MAX` boundary for `RawVec<u8>`. If there are any problems all the 32bit architectures should fail these tests.